src/test/csfle.rs

```diff
@@ -0,0 +1,1261 @@
+        let kv_client = if with_external_key_vault {
+            let mut opts = CLIENT_OPTIONS.get().await.clone();
+            opts.credential = Some(
+                Credential::builder()
+                    .username("fake-user".to_string())
+                    .password("fake-pwd".to_string())
+                    .build(),
+            );
+            Some(Client::with_options(opts)?)
+        } else {
+            None
+        };
+
+        // Setup: encrypted client.
+        let auto_enc_opts = AutoEncryptionOptions::builder()
+            .key_vault_namespace(KV_NAMESPACE.clone())
+            .key_vault_client(kv_client.clone())
+            .kms_providers(LOCAL_KMS.clone())
+            .schema_map(schema_map)
+            .extra_options(EXTRA_OPTIONS.clone())
+            .build();
+        let client_encrypted =
+            Client::with_encryption_options(CLIENT_OPTIONS.get().await.clone(), auto_enc_opts)
+                .await?;
+
+        // Setup: manual encryption.
+        let enc_opts = ClientEncryptionOptions::builder()
+            .key_vault_namespace(KV_NAMESPACE.clone())
+            .key_vault_client(kv_client.unwrap_or_else(|| client.into_client()))
+            .kms_providers(LOCAL_KMS.clone())
+            .build();
+        let client_encryption = ClientEncryption::new(enc_opts)?;
+
+        // Test: encrypted client.
+        let result = client_encrypted
+            .database("db")
+            .collection::<Document>("coll")
+            .insert_one(doc! { "encrypted": "test" }, None)
+            .await;
+        if with_external_key_vault {
+            assert!(result.is_err());
+        } else {
+            assert!(
+                result.is_ok(),
+                "unexpected error: {}",
+                result.err().unwrap()
+            );
+        }
+        // Test: manual encryption.
+        let result = client_encryption
+            .encrypt(
+                RawBson::String("test".to_string()),
+                &EncryptOptions::builder()
+                    .key(EncryptKey::Id(base64_uuid("LOCALAAAAAAAAAAAAAAAAA==")?))
+                    .algorithm(Algorithm::AeadAes256CbcHmacSha512Deterministic)
+                    .build(),
+            )
+            .await;
+        if with_external_key_vault {
+            assert!(result.is_err());
+        } else {
+            assert!(
+                result.is_ok(),
+                "unexpected error: {}",
+                result.err().unwrap()
+            );
+        }
+    }
+
+    Ok(())
+}
+
+fn load_testdata_raw(name: &str) -> Result<String> {
+    let path: PathBuf = [
+        env!("CARGO_MANIFEST_DIR"),
+        "src/test/spec/json/client-side-encryption/testdata",
+        name,
+    ]
+    .iter()
+    .collect();
+    Ok(std::fs::read_to_string(path)?)
+}
+
+fn load_testdata(name: &str) -> Result<Document> {
+    Ok(serde_json::from_str(&load_testdata_raw(name)?)?)
+}
+
+// Prose test 4. BSON Size Limits and Batch Splitting
```

Comment:
nit: for future readability it might be helpful to number the steps in this test since all of them look kind of similar at first read due to doing inserts